Generated nix-file does not build on macOS

I am trying to build an older version of vault (1.4.3) using nix-init. However, the nix file created does not build on macOS (13.3.1, aarch64):

> vendor/github.com/shirou/gopsutil/cpu/cpu_darwin_cgo.go:13:5: error: 'TARGET_OS_MAC' is not defined, evaluates to 0 [-Werror,-Wundef-prefix=TARGET_OS_]

During my research I came across this issue:

> This is because LLVM/Clang 11 automatically adds -Werror,-Wundef-prefix=TARGET_OS_ as default compiler arguments and apparently TARGET_OS_MAC isn't defined in this compilation environment.
>
> [...]
>
> I can work around this by adding #cgo CFLAGS: -Wno-undef-prefix or #include <TargetConditionals.h> to the failing import "C" comment. The latter is likely more robust and seemingly indicates that TargetConditions.h isn't being included by another #include in this block.
>
> -- https://github.com/shirou/gopsutil/issues/976

Is there a possibility that nix-init detects and fixes this problem itself? Or do I always have to adapt the files myself?

--- a/src/lang/go/deps.rs
+++ b/src/lang/go/deps.rs
@@ -1,9 +1,11 @@
-use crate::{inputs::AllInputs, macros::input_macros};
+use semver::Version;
 
-pub(super) fn load_go_dependency(inputs: &mut AllInputs, pkg: &str) {
+use crate::{inputs::AllInputs, lang::go::GoPackage, macros::input_macros};
+
+pub(super) fn load_go_dependency(inputs: &mut AllInputs, pkg: GoPackage<'_>) {
     input_macros!(inputs);
 
-    match pkg {
+    match pkg.name {
         "github.com/diamondburned/gotk4-adwaita/pkg" => build!("libadwaita"),
         "github.com/diamondburned/gotk4/pkg" => {
             native_build!("pkg-config", "wrapGAppsHook4");

--- a/src/lang/go/deps.rs
+++ b/src/lang/go/deps.rs
@@ -38,6 +38,14 @@ pub(super) fn load_go_dependency(inputs: &mut AllInputs, pkg: GoPackage<'_>) {
         }
         "github.com/itchio/ox" => framework!("Cocoa"),
         "github.com/itchio/screw" => framework!("Cocoa"),
+        "github.com/shirou/gopsutil" => {
+            if pkg.version.get().is_some_and(|version| {
+                version < Version::new(3, 21, 3)
+                    && (version.major != 2 || version < Version::new(2, 21, 11))
+            }) {
+                environ!("CGO_CFLAGS", r#""-Wno-undef-prefix""#);
+            }
+        }
         "golang.design/x/clipboard" => {
             build!("xorg.libX11"; linux);
             framework!("Cocoa");

--- a/src/lang/go/mod.rs
+++ b/src/lang/go/mod.rs
@@ -4,15 +4,47 @@ mod goreleaser;
 use std::{
     fs::File,
     io::{BufRead, BufReader},
+    str::SplitWhitespace,
 };
 
-pub use goreleaser::write_ldflags;
+use semver::Version;
 
+pub use self::goreleaser::write_ldflags;
 use crate::{inputs::AllInputs, lang::go::deps::load_go_dependency};
+
+struct GoPackage<'a> {
+    name: &'a str,
+    version: GoVersion<'a>,
+}
+
+struct GoVersion<'a> {
+    line: SplitWhitespace<'a>,
+}
+
+impl<'a> GoPackage<'a> {
+    fn from_line(mut line: SplitWhitespace<'a>) -> Option<Self> {
+        Some(GoPackage {
+            name: line.next()?,
+            version: GoVersion { line },
+        })
+    }
+}
+
+impl<'a> GoVersion<'a> {
+    fn get(mut self) -> Option<Version> {
+        self.line
+            .next()?
+            .strip_prefix('v')?
+            .split('/')
+            .next()?
+            .parse()
+            .ok()
+    }
+}
 
 pub fn load_go_dependencies(inputs: &mut AllInputs, go_sum: &File) {
     for line in BufReader::new(go_sum).lines().map_while(Result::ok) {
-        if let Some(pkg) = line.split_whitespace().next() {
+        if let Some(pkg) = GoPackage::from_line(line.split_whitespace()) {
             load_go_dependency(inputs, pkg);
         };
     }
@@ -23,8 +55,12 @@ mod tests {
     use std::fs::File;
 
     use insta::assert_debug_snapshot;
+    use semver::Version;
 
-    use crate::{inputs::AllInputs, lang::go::load_go_dependencies};
+    use crate::{
+        inputs::AllInputs,
+        lang::go::{load_go_dependencies, GoPackage},
+    };
 
     #[test]
     fn basic() {
@@ -34,5 +70,23 @@ mod tests {
             &File::open("src/lang/go/fixtures/basic/go.sum").unwrap(),
         );
         assert_debug_snapshot!(inputs);
+    }
+
+    #[test]
+    fn parse() {
+        assert_debug_snapshot!(parse_line(
+            "golang.org/x/mod v0.11.0 h1:bUO06HqtnRcc/7l71XBe4WcqTZ+3AH1J59zWDDwLKgU="
+        ));
+        assert_debug_snapshot!(parse_line(
+            "github.com/spf13/cobra v1.7.0/go.mod h1:uLxZILRyS/50WlhOIKD7W6V5bgeIt+4sICxh6uRMrb0="
+        ));
+        assert_debug_snapshot!(parse_line(
+            "golang.org/x/crypto v0.0.0-20221012134737-56aed061732a/go.mod h1:IxCIyHEi3zRg3s0A5j5BB6A9Jmi73HwBIUl50j+osU4="
+        ));
+    }
+
+    fn parse_line(line: &str) -> (&str, Version) {
+        let pkg = GoPackage::from_line(line.split_whitespace()).unwrap();
+        (pkg.name, pkg.version.get().unwrap())
     }
 }

--- a/src/macros.rs
+++ b/src/macros.rs
@@ -21,14 +21,14 @@ macro_rules! input_macros {
         #[allow(unused_macros)]
         macro_rules! environ {
             ($name:expr, $value:expr) => {
-                environ!($name, $value;);
+                environ!($name, $value;)
             };
-            ($name:expr, $value:expr; $_($tt:tt)*) => {
+            ($name:expr, $value:expr; $_($tt:tt)*) => {{
                 $inputs.env.insert(
                     $name.into(),
                     ($value.into(), vec![$_($tt)*]),
                 );
-            };
+            }};
         }
 
         // native build inputs

